### PR TITLE
Implement reshapeable CTCGreedyDecoderPlusSparseToDense transformation and test

### PR DESCRIPTION
**Pull request by rkazants**
Description: Old transformation is not reshapeable and has limitation that requires to have seq_len value for IR generation

JIRA: 31610 and 37580

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests: N/A
* [x]  Transformation tests
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py N/A
* [x]  Model Optimizer IR Reader check. Checked

Documentation:
* [x]  Supported frameworks operations list. N/A - nothing changed in terms of new framework operations
* [x]  Supported **public** models list. N/A - it relates to already enabled model.
* [x]  New operations specification. N/A - nothing changed in terms of new operations for opset
* [x]  Guide on how to convert the **public** model - nothing changed in MO cmd
* [x]  User guide update. N/A

Signed-off-by: Roman Kazantsev <roman.kazantsev@intel.com>

### CODE REVIEW

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py [hunk 1]
```diff
@@ -29,52 +35,115 @@ class CTCGreedyDecoderReplacement(FrontReplacementSubgraph):
...
+
     @staticmethod
     def pattern(**kwargs):
         return dict(
             nodes=[
+                ('transpose', dict(op='Transpose')),
+                ('shape', dict(op='ShapeOf')),
+                ('shape_1', dict(op='ShapeOf')),
+                ('strided_slice', dict(op='StridedSlice')),
+                ('stack', dict(op='Const', value=[1])),
```
Comment (jane-intel) [inline model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py:55]: it is not safe to match node with values like that. we should check like 
`value=lambda v: v is not None and np.array_equal(v, [1])`

**Review by jane-intel — COMMENTED** [review]
How many models are affected? Does TF generate the same sub-graph (from the pattern) all the time?

**Comment by rkazants** [comment]
> How many models are affected? Does TF generate the same sub-graph (from the pattern) all the time?
We have a couple of models with CTCGreedyDecoder for which separate transformation is implemented. Now we can't propose generic pattern because CTCGreedyDecoder generate output in sparse tensor that means dynamic size for values that affects next layers and also opset CTCGreedyDecoder has different format for `seq_len` input from TF's one.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py (hunk 1 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py:55]: Ok, will fix it.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py [hunk 2]
```diff
@@ -29,52 +35,115 @@ class CTCGreedyDecoderReplacement(FrontReplacementSubgraph):
...
+        # create CTCGreedyDecoder with the sequence mask instead of sequence length
+        ctc_greedy_decoder.in_port(1).disconnect()
+        ctc_greedy_decoder.in_port(1).connect(sequence_mask.out_port(0))
+
+        # remove fill and pack nodes since they are now in unconnected component
+        graph.remove_nodes_from([fill.id, dims.id])
+
+        # transform opset CTCGreedyDecoder output to TensorFlow's one that has a shape [N, T]
+        # opset CTCGreedyDecoder has an output with a shape [N, T, 1, 1]
+        squeeze_dec_seq = create_op_with_const_inputs(graph, Squeeze, {1: int64_array([2, 3])},
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py:133]: This operation is a Squeeze from 4D to 2D and the MO will insert Transpose from NHWC to NCHW. Could you attach the screenshot of the part of the IR model with the sub-graph around the CTCGreedyDecoder? We may need to mark the Squeeze node with the `mark_input_as_in_correct_layout` function.

**Comment by lazarevevgeny** [comment]
> > How many models are affected? Does TF generate the same sub-graph (from the pattern) all the time?
> 
> @jane-intel , we have a couple of models with CTCGreedyDecoder for which separate transformation is implemented. Now we can't propose generic pattern because CTCGreedyDecoder generate output in sparse tensor that means dynamic size for values that affect next layers and also opset CTCGreedyDecoder has `seq_len` input format different from TF's one.

The pattern is really big and it contains Transpose operation. Is it really common for all TF models where GD is generated? Can we have a layer test where the GD is created and the same model sub-graph is generated? Before your change the pattern was way smaller.

**Comment by rkazants** [comment]
> > How many models are affected? Does TF generate the same sub-graph (from the pattern) all the time?
> 
> @jane-intel , we have a couple of models with CTCGreedyDecoder for which separate transformation is implemented. Now we can't propose generic pattern because CTCGreedyDecoder generate output in sparse tensor that means dynamic size for values that affect next layers and also opset CTCGreedyDecoder has `seq_len` input format different from TF's one.

I think we have not too many that have such pattern or even this one is only. To have such pattern is a cost to have reshapeable transformation. The previous implementation was more generic but not reshapeable. For this model we are lucky that `seq_len` is not a model parameter and computed from shape of logits on intermediate step.

In order to implement more generic pattern (CTCGreedyDecoder+SparseToDense) we have to implement new CTCGreedyDecoder with TF seq_len format or an operation SeqLen2SeqMask.

**Comment by rkazants** [comment]
> > > How many models are affected? Does TF generate the same sub-graph (from the pattern) all the time?
> > 
> > 
> > @jane-intel , we have a couple of models with CTCGreedyDecoder for which separate transformation is implemented. Now we can't propose generic pattern because CTCGreedyDecoder generate output in sparse tensor that means dynamic size for values that affect next layers and also opset CTCGreedyDecoder has `seq_len` input format different from TF's one.
> 
> The pattern is really big and it contains Transpose operation. Is it really common for all TF models where GD is generated? Can we have a layer test where the GD is created and the same model sub-graph is generated? Before your change the pattern was way smaller.

@lazarevevgeny, I think we have not too many that have such pattern or even this one is only. To have such pattern is a cost to have reshapeable transformation. The previous implementation was more generic but not reshapeable. For this model we are lucky that `seq_len` is not a model parameter and computed from shape of logits on intermediate step.

In order to implement more generic pattern (CTCGreedyDecoder+SparseToDense) we have to implement new CTCGreedyDecoder with TF seq_len format or an operation SeqLen2SeqMask.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py (hunk 2 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py:133]: image

Comment (rkazants) [reply model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py:133]: @lazarevevgeny , It seems no need to set this attribute. The screenshot is attached.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py (hunk 1 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py]: Fixed.

**Comment by rkazants** [comment]
As we agreed in the meeting I have added more generic CTCGreedyDecoder transformation that is not reshape-able. Also it adds middle-replacer for transformation sequence length to a mask and tests.
I checked this transformation when the reshapeable transformation is switched off.
@lazarevevgeny , @jane-intel , please review.

File: model-optimizer/extensions/middle/SequenceLengthToMask.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+        # transform a sequence length to a sequence mask
+        logits_shape = ctc_greedy_decoder.in_port(0).data.get_shape()
+        assert logits_shape is not None and len(logits_shape) == 3, \
+            "Incorrect shape for logits input of {} node".format(ctc_greedy_decoder_name)
+        batch_size = logits_shape[1]
+        time_size = logits_shape[0]
+        mask_value = np.zeros([batch_size, time_size], dtype=np.float)
+        for sample_ind in range(batch_size):
+            sample_seq_length = sequence_length_value[sample_ind]
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/SequenceLengthToMask.py:61]: Change to:
```
        for sample_ind, sample_seq_length in enumerate(sequence_length_value):
```

File: model-optimizer/extensions/middle/SequenceLengthToMask.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            sample_mask_value = np.zeros([time_size])
+            sample_mask_value[0:sample_seq_length] = 1
+            mask_value[sample_ind, :] = sample_mask_value
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/SequenceLengthToMask.py:64]: Can we rewrite these 3 lines with: `mask_value[sample_ind, 0:sample_seq_length] = 1`?

File: model-optimizer/extensions/middle/SequenceLengthToMask.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        mask_value = np.transpose(mask_value)
+
+        # create Const node with computed mask value
+        mask_node = Const(graph, {'name': ctc_greedy_decoder_name + '/Mask',
+                                  'value': mask_value}).create_node()
+
+        # connect computed mask to CTCGreedyDecoder node
+        ctc_greedy_decoder.in_port(1).disconnect()
+        ctc_greedy_decoder.in_port(1).connect(mask_node.out_port(0))
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/SequenceLengthToMask.py:73]: `ctc_greedy_decoder.in_port(1).get_connection().set_source(mask_node.out_port(0))`

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py [hunk 6]
```diff
@@ -14,67 +14,223 @@
...
+        # TODO: Once Inference Engine's CTCGreedyDecoder starts to support sequence length format like in TensorFlow,
+        # CTCGreedyDecoderReplacement2 needs to be removed and CTCGreedyDecoderReplacement, a more generic
+        # transformation, needs to be adopted for all cases
+        ctc_greedy_decoder = match['decoder']
+        cast = match['cast']
+        sparse_to_dense = match['sparse_to_dense']
+        sparse_to_dense_name = sparse_to_dense.soft_get('name', sparse_to_dense.id)
+
+        # remove edges from CTCGreedyDecoder node since top nodes willbe changed
+        graph.remove_edge(ctc_greedy_decoder.id, sparse_to_dense.id)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py:76]: Can we replace these function calls with methods like `node.in_port(XXX).disconnect()`?

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/SequenceLengthToMask.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        return dict(
+            nodes=[
+                ('ctc_greedy_decoder', dict(op='CTCGreedyDecoder',
+                                            use_mask_format=True)),
+            ],
+            edges=[
+            ])
+
+    def replace_pattern(self, graph: Graph, match: dict):
+        ctc_greedy_decoder = match['ctc_greedy_decoder']
```
[the remark this answers is not in the document]
Comment (rkazants) [reply model-optimizer/extensions/middle/SequenceLengthToMask.py]: Fixed.

File: model-optimizer/extensions/middle/SequenceLengthToMask.py (hunk 3 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/SequenceLengthToMask.py]: Fixed.

File: model-optimizer/extensions/middle/SequenceLengthToMask.py (hunk 4 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/SequenceLengthToMask.py]: Yes, it will be much better. Thanks.
Fixed.

File: model-optimizer/extensions/middle/SequenceLengthToMask.py (hunk 5 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/SequenceLengthToMask.py]: Fixed.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py (hunk 6 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py]: Yes, fixed.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py [hunk 8]
```diff
@@ -14,67 +14,223 @@
...
+        graph.remove_nodes_from([sparse_to_dense.id, cast.id])
+
+        # mark CTCGreedyDecoder node as a node that requires transformation of sequence length to a mask format
+        # in the middle phase
+        ctc_greedy_decoder['use_mask_format'] = True
+
+        # unless the second input of CTCGreedyDecoder is a parameter, it enforces MO to use --static-shape
+        # to try getting the second input with a value
+        sequence_length_node = ctc_greedy_decoder.in_node(1)
+        if sequence_length_node.op != 'Parameter' and not graph.graph['cmd_params'].static_shape:
```
Comment (jane-intel) [inline model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py:106]: Proper way to get such an info is to start a search from the `ctc_greedy_decoder.in_port(1)` and find one of the three cases:
- constant or constant sub-graph that depends on Const layers only -- all good
- constant sub-graph that depend on ShapeOfs -- should use static_shape (there are Size and Rank operations that should be taken into account here)
- data sub-graph that depend on Parameter -- should raise another message

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement_test.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+                                 ('sequence_mask', 'decoder', {'out': 0, 'in': 1}),
+                                 ('decoder', 'squeeze_dec_seq', {'out': 0, 'in': 0}),
+                                 ('squeeze_axes', 'squeeze_dec_seq', {'out': 0, 'in': 1}),
+                                 ('squeeze_dec_seq', 'cast_to_int', {'out': 0, 'in': 0}),
+                                 ('cast_to_int', 'last', {'out': 0, 'in': 0}),
+                                 ],
+                                nodes_with_edges_only=True)
+
+        (flag, resp) = compare_graphs(graph, graph_ref, 'last', check_op_attrs=True)
+        self.assertTrue(flag, resp)
```
Comment (jane-intel) [inline model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement_test.py:159]: you care about node names in the transformations -- please check it in the test

**Review by jane-intel — COMMENTED** [review]
If my suggestion with a search is an overkill for such a task -- lets forget about it.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py (hunk 8 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py:106]: I think that is not so important now since we have a plan to re-design `CTCGreedyDecoder` and there is no such model. Let us cancel this.

**Comment by rkazants** [comment]
> If my suggestion with a search is an overkill for such a task -- lets forget about it.

Yes, it is an overkill for now. Let us cancel it.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py (hunk 8 above)
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py:106]: Please, change `sequence_length_node.op` to `sequence_length_node.soft_get('op')`. It is a bit more reliable.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py [hunk 10]
```diff
@@ -14,67 +14,223 @@
...
+        # preserve output name from original graph
+        rename_nodes([(sparse_to_dense, sparse_to_dense_name + '/AbandonedName'),
+                      (cast_to_int, sparse_to_dense_name)])
+
+        # set output of the new sub-graph as a source for SparseToDense consumer
+        sparse_to_dense.out_port(0).get_connection().set_source(cast_to_int.out_port(0))
+
+        # cleanup a graph: disconnect CTCGreedyDecoder from Cast and SparseToDense
+        # since opset CTCGreedyDecoder outputs decoded sequence in a dense format
+        cast.in_port(0).disconnect()
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py:233]: I doubt that it is necessary to disconnect these ports since we remove these nodes 3 lines below. So, please, remove these lines.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement_test.py (hunk 9 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement_test.py:163]: Fixed

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py (hunk 8 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py]: Fixed.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py (hunk 10 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py]: Fixed.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement_test.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+                                 ('seq_len', 'decoder', {'out': 0, 'in': 1}),
+                                 ('decoder', 'squeeze_dec_seq', {'out': 0, 'in': 0}),
+                                 ('squeeze_axes', 'squeeze_dec_seq', {'out': 0, 'in': 1}),
+                                 ('squeeze_dec_seq', 'cast_to_int', {'out': 0, 'in': 0}),
+                                 ('cast_to_int', 'last', {'out': 0, 'in': 0}),
+                                 ],
+                                nodes_with_edges_only=True)
+
+        (flag, resp) = compare_graphs(graph, graph_ref, 'last', check_op_attrs=True)
+        self.assertEqual(graph.get_op_nodes(op='Cast')[0]['name'], 'sparse_to_dense',
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement_test.py:66]: `self.assertTrue(len(graph.get_op_nodes(op='Cast')) == 1 and graph.get_op_nodes(op='Cast')[0]['name'] == 'sparse_to_dense'`. The same comment to another unit test.

**Review by lazarevevgeny — APPROVED** [review]
Approved. But, please, make a small change in the unit test before merge.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement_test.py (hunk 11 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement_test.py]: Fixed.

**Comment by rkazants** [comment]
> Approved. But, please, make a small change in the unit test before merge.

Done.

**Review by jane-intel — APPROVED** [review]
